### PR TITLE
Improve crystal syntaxes

### PR DESCRIPTION
**Pull request by icyleaf**
What's new:

- Sync vscode-ruby (Feb 2018)
  - Matches function call
  - Changes `require` and constant style of color
  - Category keywords make it clearly
  - Minor bugfix and optimizing
- Matches `[]?`, `[]=` and `^` method name
- Matches `__DIR__`
- Matches  annotation (#68)
- Adds Folding with record, annotation
- Removes unused rule: deprecated `else if`
- Highlight `type` function in `Lib` only (#70)

crystal_syntaxes

BTW, append "information_for_contributors" and "version" key in crystal.json.

### CODE REVIEW

**Comment by faustinoaq** [comment]
@icyleaf Thank you!

**Comment by icyleaf** [comment]
I fixed conflicts :)

**Comment by faustinoaq** [comment]
> The `"comment"` style is inconsistent and varies from leading upper casing over leading lower casing to space leat lower casing.

Hi @fibric , what do you mean? :point_up: I think vscode manages comments inside extension itself, I'm merging this, Also I'm testing a new syntax (based on original texmate bundle)  :sweat_smile:

File: syntaxes/crystal.json [hunk 1]
```diff
@@ -105,24 +112,33 @@
...
+					"name": "entity.name.lib.type.crystal"
+				},
+				"3": {
+					"name": "keyword.control.lib.crystal"
+				},
+				"4": {
+					"name": "entity.name.lib.type.value.crystal"
+				}
+			},
+			"comment": "type in lib",
```
Comment (fibric) [inline syntaxes/crystal.json]: comment style: lower case

File: syntaxes/crystal.json [hunk 2]
```diff
@@ -105,24 +112,33 @@
...
+			"match": "(?<!\\.)\\b(type)\\s+([A-Z]\\w+)\\s*(=)\\s*(.+)",
+			"name": "meta.lib.type.crystal"
 		},
 		{
 			"comment": "everything being a reserved word, not a value and needing a 'end' is a..",
```
Comment (fibric) [inline syntaxes/crystal.json]: comment style: lower case

File: syntaxes/crystal.json [hunk 3]
```diff
@@ -105,24 +112,33 @@
...
-			"match": "(?<!\\.)\\b(alias|as|begin|case|abstract|class|ensure|for|fun|if|in|lib|module|of|out|previous_def|private|protected|rescue|struct|with|union|enum|macro|then|type|unless|until|while|select)\\b(?![?!])",
-			"name": "keyword.control.primary.crystal"
+			"match": "(?<!\\.)\\b(fun|begin|case|class|else|elsif|end|ensure|for|if|in|module|rescue|then|unless|until|when|while)\\b(?![?!])",
+			"name": "keyword.control.crystal"
 		},
 		{
 			"comment": "everything being a reserved word, not a value and needing a 'end' is a..",
```
Comment (fibric) [inline syntaxes/crystal.json]: comment style: lower case

File: syntaxes/crystal.json [hunk 4]
```diff
@@ -105,24 +112,33 @@
...
-		},
-		{
-			"comment": "Give the end keyword an additional scope",
-			"match": "(?<!\\.)\\b(end)\\b(?![?!])",
-			"name": "keyword.control.secondary.end.crystal"
+			"match": "(?<!\\.)\\b(as|abstract|ensure|fun|in|lib|of|out|previous_def|private|protected|struct|with|union|enum|macro|select|record)\\b(?![?!])",
+			"name": "keyword.control.primary.crystal"
 		},
 		{
 			"comment": "contextual smart pair support for block parameters",
```
Comment (fibric) [inline syntaxes/crystal.json]: comment style: lower case

File: syntaxes/crystal.json [hunk 5]
```diff
@@ -135,33 +151,36 @@
 			"name": "meta.syntax.crystal.start-block"
 		},
 		{
-			"comment": " as above, just doesn't need a 'end' and does a logic operation",
-			"match": "(?<!\\.)\\b(and|not|or)\\b",
-			"name": "keyword.operator.logical.crystal"
-		},
-		{
-			"comment": " just as above but being not a logical operation",
+			"comment": "Just as above but being not a logical operation",
```
Comment (fibric) [inline syntaxes/crystal.json]: comment style: upper case

File: syntaxes/crystal.json [hunk 6]
```diff
@@ -135,33 +151,36 @@
...
 		},
 		{
-			"comment": " everything being a method but having a special function is a..",
-			"match": "\\b(initialize|new|loop|include|extend|raise|getter|setter|property|class_getter|class_setter|class_property|describe|it|with|delegate|def_hash|def_equals|def_equals_and_hash|forward_missing_to|record|assert_responds_to|spawn)\\b[!?]?",
-			"name": "keyword.control.special-method.crystal"
+			"match": "\\bself\\b(?![?!])",
+			"name": "variable.language.self.crystal"
+		},
+		{
+			"comment": "Everything being a method but having a special function is a..",
```
Comment (fibric) [inline syntaxes/crystal.json]: comment style: upper case

File: syntaxes/crystal.json [hunk 7]
```diff
@@ -232,16 +251,23 @@
 				}
 			]
 		},
+		{
+			"match": "\\b[_A-Z]+\\b",
+			"name": "variable.other.constant.crystal",
+			"comment": "Constant name in any where"
```
Comment (fibric) [inline syntaxes/crystal.json]: comment style: upper case

File: syntaxes/crystal.json [hunk 8]
```diff
@@ -232,16 +251,23 @@
...
+		},
 		{
 			"match": "\\b[A-Z\\x{80}-\\x{10FFFF}][\\w\\x{80}-\\x{10FFFF}]*",
-			"name": "support.class.crystal"
+			"name": "support.class.crystal",
+			"commenct": "Literals name of Crystal"
```
Comment (fibric) [inline syntaxes/crystal.json]: comment style: upper case

File: syntaxes/crystal.json [hunk 9]
```diff
@@ -232,16 +251,23 @@
...
 		},
 		{
-			"match": "\\b[A-Z\\x{80}-\\x{10FFFF}][\\w\\x{80}-\\x{10FFFF}]*\\b",
-			"name": "variable.other.constant.crystal"
+			"match": "\\b(abort|at_exit|caller|delay|exit|fork|future|lazy|loop|main|p|pp|print|printf|puts|rand|read_line|sleep|spawn|sprintf|system|with_color|debugger)\\b[!?]?",
+			"name": "support.function.kernel.crystal",
+			"comment": "Fetch from https://crystal-lang.org/api/0.27.0/toplevel.html"
```
Comment (fibric) [inline syntaxes/crystal.json]: comment style: upper case

File: syntaxes/crystal.json [hunk 10]
```diff
@@ -253,18 +279,40 @@
 					"name": "punctuation.definition.parameters.crystal"
 				}
 			},
-			"comment": "Parentesis on Crystal are mandatory. The method pattern comes from the symbol pattern, see there for a explaination",
-			"contentName": "variable.parameter.function.crystal",
-			"end": "\\)\\s*$|\\)\\s*:|\\)\\s*;",
-			"endCaptures": [
-				{
+			"comment": "The method pattern comes from the symbol pattern. See there for an explanation.",
```
Comment (fibric) [inline syntaxes/crystal.json]: comment style: upper case

File: syntaxes/crystal.json [hunk 11]
```diff
@@ -277,23 +325,29 @@
 					"name": "entity.name.function.crystal"
 				}
 			},
-			"comment": "The optional name is just to catch the def also without a method-name",
-			"match": "(?x) (?=def\\b)(?<=^|\\s)(def|fun)\\b( \\s+( (?>[a-zA-Z_\\x{80}-\\x{10FFFF}][\\w\\x{80}-\\x{10FFFF}]*(?>\\.|::))?(?>[a-zA-Z_\\x{80}-\\x{10FFFF}][\\w\\x{80}-\\x{10FFFF}]*(?>[?!]|=(?!>))?|===?|>[>=]?|<=>|<[<=]?|[%&`\/\\|]|\\*\\*?|=?~|[-+]@?|\\[\\]=?) ) )?",
+			"comment": " the optional name is just to catch the def also without a method-name",
```
Comment (fibric) [inline syntaxes/crystal.json]: comment style: leading space followed by lower case

File: syntaxes/crystal.json [hunk 12]
```diff
@@ -277,23 +325,29 @@
...
-			"captures": [
-				{
-					"name": "punctuation.definition.constant.crystal"
+			"beginCaptures": {
+				"0": {
+					"name": "punctuation.definition.symbol.begin.crystal"
 				}
-			],
+			},
+			"comment": "symbol literal with '' delimitor",
```
Comment (fibric) [inline syntaxes/crystal.json]: comment style: lower case

File: syntaxes/crystal.json [hunk 13]
```diff
@@ -303,13 +357,19 @@
...
-			"captures": [
-				{
-					"name": "punctuation.definition.constant.crystal"
+			"beginCaptures": {
+				"0": {
+					"name": "punctuation.section.symbol.begin.crystal"
 				}
-			],
+			},
+			"comment": "symbol literal with \"\" delimitor",
```
Comment (fibric) [inline syntaxes/crystal.json]: comment style: lower case

File: syntaxes/crystal.json [hunk 14]
```diff
@@ -321,23 +381,23 @@
 		},
 		{
 			"comment": "Needs higher precidence than regular expressions.",
```
Comment (fibric) [inline syntaxes/crystal.json]: comment style: upper case

File: syntaxes/crystal.json [hunk 15]
```diff
@@ -321,23 +381,23 @@
...
-			"beginCaptures": [
-				{
+			"beginCaptures": {
+				"0": {
 					"name": "punctuation.definition.string.begin.crystal"
 				}
-			],
-			"comment": "single quoted string (does not allow interpolation)",
+			},
+			"comment": "string literal with '' delimitor",
```
Comment (fibric) [inline syntaxes/crystal.json]: comment style: lower case

File: syntaxes/crystal.json [hunk 16]
```diff
@@ -348,19 +408,19 @@
...
-			"beginCaptures": [
-				{
+			"beginCaptures": {
+				"0": {
 					"name": "punctuation.definition.string.begin.crystal"
 				}
-			],
-			"comment": "double quoted string (allows for interpolation)",
+			},
+			"comment": "string literal with interpolation and \"\" delimitor",
```
Comment (fibric) [inline syntaxes/crystal.json]: comment style: lower case

File: syntaxes/crystal.json [hunk 17]
```diff
@@ -372,18 +432,18 @@
...
 			"begin": "`",
-			"beginCaptures": [
-				{
+			"beginCaptures": {
+				"0": {
 					"name": "punctuation.definition.string.begin.crystal"
 				}
-			],
+			},
 			"comment": "execute string (allows for interpolation)",
```
Comment (fibric) [inline syntaxes/crystal.json]: comment style: lower case

File: syntaxes/crystal.json [hunk 18]
```diff
@@ -1299,21 +1362,38 @@
...
+		{
+			"match": "(?<=^|[ \\t])!|&&|\\|\\||\\^",
 			"name": "keyword.operator.logical.crystal"
 		},
 		{
 			"match": "(\\{\\%|\\%\\}|\\{\\{|\\}\\})",
 			"name": "keyword.operator.macro.crystal"
 		},
+		{
+			"comment": "Safe navigation operator",
```
Comment (fibric) [inline syntaxes/crystal.json]: comment style: upper case

File: syntaxes/crystal.json [hunk 19]
```diff
@@ -1353,6 +1433,31 @@
...
+					"match": "([a-zA-Z0-9_!?]+)(?=\\()",
+					"name": "entity.name.function.crystal"
+				},
+				{
+					"include": "$self"
+				}
+			]
+		},
+		{
+			"comment": "This is kindof experimental. There really is no way to perfectly match all regular variables, but you can pretty well assume that any normal word in certain curcumstances that havnt already been scoped as something else are probably variables, and the advantages beat the potential errors",
```
Comment (fibric) [inline syntaxes/crystal.json]: comment style: upper case

**Comment by fibric** [comment]
@faustinoaq I addressed the `comment` property, and it's usage across the file.

**Comment by faustinoaq** [comment]
@fibric Thank you I'll change this on a new PR :+1:

**Pull request merged**